opentracing/src/main/java/opentracing/SpanPropagator.java

```diff
@@ -0,0 +1,53 @@
+/**
+ * Copyright 2016 The OpenTracing Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package opentracing;
+
+import java.util.Map;
+
+/**
+ * Encodes or Decodes a {@linkplain Span} in binary or text formats.
+ *
+ * <p>The toXXX methods are expected to serialize spans into a pair of values representing
+ * separately the span / span identity, and the trace attributes. This is done specifically
+ * for binary protocols that may represent tracing identity in a dedicated fixed-length slot in the
+ * binary message format, so that it can be inspected efficiently by the middleware / routing layers
+ * without parsing the whole message.
+ */
+public interface SpanPropagator {
```

Comment:
For example

    public interface SpanExtractor {
        Span getSpan();
    }

Again the notion of separation between span identity and trace attributes is just not needed in the end-user API.

If the implementation wants to provide optimisations at the wire level (something that should not be polluting end-user API) then this can be done inside the implementation of the Span that's returned from the SpanExtractor.

At the end of the day it's neither the middleware framework nor the OT implementation that determines where the barrier between span identity and trace attributes lies, but the wire protocol itself.

I'm also not sure what type of optimisation such a routing layer could provide anyway, will it know the length of bytes/characters in the trace attribute section and be able to skip over it? Will it have access to this level in the protocol middleware to do such skipping? Or will the common implementation be in fact still to read through these bytes/characters but just not put them back into wire? If it's the latter it's only a very small amount of cpu optimisation that's relevant here so sounds like a case of pre-mature optimisation. 

For the situation of adding trace attributes at a proxy routing layer it all sounds a little flakey to be playing around with bytes in this manner; the presumption that bytes can simply be tacked on to the end of an existing byte stream without having read it. And it's obvious you can't do that with Map<String,String>.